saml/saml.php

```diff
@@ -0,0 +1,429 @@
+<?php
+/*
+ * Name: SAML SSO and SLO
+ * Description: replace login and registration with a SAML identity provider.
+ * Version: 0.0
+ * Author: Ryan <https://friendica.verya.pe/profile/ryan>
+ */
+use Friendica\Core\Hook;
+use Friendica\Core\Logger;
+use Friendica\Core\Renderer;
+use Friendica\Core\Session;
+use Friendica\Database\DBA;
+use Friendica\DI;
+use Friendica\Model\User;
+use Friendica\Util\Strings;
+use OneLogin\Saml2\Auth;
+
+require_once(__DIR__ . '/vendor/autoload.php');
+
+define("PW_LEN", 32); // number of characters to use for random passwords
+
+function saml_module($a) {}
+
+function saml_init($a) {
+    if ($a->argc < 2) return;
+
+    switch ($a->argv[1]) {
+        case "metadata.xml":
+            saml_metadata();
+            break;
+        case "sso":
+            saml_sso_reply($a);
+            break;
+        case "slo":
+            saml_slo_reply();
+            break;
+	case "moo":
+	    echo DI::baseUrl();
+	    echo $_SERVER['REQUEST_URI'];
+	    break;
+    }
+    exit();
+}
+
+function saml_metadata() {
+    try {
+        $settings = new \OneLogin\Saml2\Settings(saml_settings());
+        $metadata = $settings->getSPMetadata();
+        $errors = $settings->validateMetadata($metadata);
+
+        if (empty($errors)) {
+            header('Content-Type: text/xml');
+            echo $metadata;
+        } else {
+            throw new \OneLogin\Saml2\Error(
+                'Invalid SP metadata: '.implode(', ', $errors),
+                \OneLogin\Saml2\Error::METADATA_SP_INVALID
+            );
+        }
+    } catch (Exception $e) {
+	Logger::error($e->getMessage());
+    }
+}
+
+function saml_install() {
+    Hook::register('login_hook', 'addon/saml/saml.php', 'saml_sso_initiate');
+    Hook::register('logging_out', 'addon/saml/saml.php', 'saml_slo_initiate');
+    Hook::register('head', 'addon/saml/saml.php', 'saml_head');
+    Hook::register('footer', 'addon/saml/saml.php', 'saml_footer');
```

Comment:
Portability: This code is correct, but we've started using `__FILE__` instead to improve code reuse between addons and prevent rare issues with moved files.
```suggestion
    Hook::register('login_hook', __FILE__, 'saml_sso_initiate');
    Hook::register('logging_out', __FILE__, 'saml_slo_initiate');
    Hook::register('head', __FILE__, 'saml_head');
    Hook::register('footer', __FILE__, 'saml_footer');
```